### PR TITLE
Compose: Rework types of createHigherOrderComponent for closer match to API

### PR DESCRIPTION
**Pull request by dmsnell**
## Status

**Ready for final review ✓**

There are a couple things I'm unhappy with about this refactor. I think it's more accurate
than the existing types, which I find confusing and think might be _wrong_. Still, it doesn't
handle components with `ref` parameters well (see `with-global-events`) and I would
prefer that we have full inference with fewer necessary type annotations.

**Update**: For now I want to ignore the `ref` issues and `with-global-events` since that
component has a number of more difficult type issues _and_ it's marked as _deprecated_.
Let's not hold up this work to spend time describing a function we don't want people using.

<details closed>
<summary>Type annotations we don't want</summary>
Everything is passing the tests right now though so I don't want to totally hold off on this,
but I'd like to let it stew for a bit and see if anyone has suggestions on how to close the gap.
For reference, here is how I think `withInstanceId` should be able to be written (pay attention
to the lack of type annotations - we shouldn't be losing any auto-complete or type safety).

```ts
const withInstanceId = createHigherOrderComponent< {
	instanceId: string | number;
} >( WrappedComponent ) => {
	return ( props ) => {
		const instanceId = useInstanceId( WrappedComponent );
			return (
				<WrappedComponent
					{ ...props }
					instanceId={ instanceId }
				/>
			);
		};
	},
	'withInstanceId'
);
```
</details>

## Description

The existing types in `createHigherOrderComponent<TOuter, TInner>` are somewhat
_surprising_ because they require defining what possible prop types are allowed
for components wrapped by the higher order component it produces. Additionally
it doesn't concern itself with preventing the wrapped component from accepting
or demanding props it itself provides.

In this patch we're reworking the types to only require a single type parameter,
which is the type of props that the higher order component injects into the
wrapped component. The prop types for the wrapped component are inferred only
when wrapping a component, and the injected props are removed from the external
interface to avoid confusion.

## How has this been tested?
This is a type-only change that shouldn't impact the generated bundles.

## Types of changes
Updating type signature for `createHigherOrderComponent`

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). 
- [x] I've updated related schemas if appropriate.

cc: @sirbrillig

### CODE REVIEW

File: packages/compose/src/higher-order/if-condition/index.tsx
Comment (zaguiini) [inline packages/compose/src/higher-order/if-condition/index.tsx]: Instead of telling ESLint to shut up, how about we export that type from our package instead?

File: packages/compose/src/higher-order/with-safe-timeout/index.tsx
Comment (zaguiini) [inline packages/compose/src/higher-order/with-safe-timeout/index.tsx:73]: Why do we need this type assertion?

File: packages/compose/src/higher-order/if-condition/index.tsx
Comment (dmsnell) [reply packages/compose/src/higher-order/if-condition/index.tsx]: Not really telling it to "shut up" 😆 

There are almost ninety other instances of disabling this error for a type import. I'm happy to try and get back to adjust `@wordpress/element` but it's still a JavaScript file without exported types. I'd prefer to not hold up this PR while waiting to change `@wordpress/element` and update all of those other existing cases.

Thoughts?

File: packages/compose/src/higher-order/with-safe-timeout/index.tsx
Comment (dmsnell) [reply packages/compose/src/higher-order/with-safe-timeout/index.tsx:73]: I'm not totally confident on my understand here, but I believe that it's because of the `Omit`. What I'm trying to convey is that the wrapped component doesn't want the timer functions to be passed in from the outside, but it still provides them to the wrapped component. I think there could be an issue here with React saying that `props` could contain any number of properties and TypeScript wants to complain that we might get ones we don't want or expect.

File: packages/compose/src/higher-order/if-condition/index.tsx
Comment (gziolo) [reply packages/compose/src/higher-order/if-condition/index.tsx]: It's disabling this rule:
https://github.com/WordPress/gutenberg/blob/cc569b28ab7612b0bd530cc53872677374553750/.eslintrc.js#L84-L88

It was discussed several times before. Ideally, we would re-import APIs we use from `@wordpress/element` for consistency with how React is used. However, it never seemed to be a simple task with TypeScript. People started disabling this rule when importing rules. We should instead seek ways to update this rule so it doesn't error for types.

Comment (gziolo) [reply packages/compose/src/higher-order/if-condition/index.tsx]: It looks like it's a simple as using a different rule for TS files:

https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-restricted-imports.md

Comment (gziolo) [reply packages/compose/src/higher-order/if-condition/index.tsx]: PR is ready: https://github.com/WordPress/gutenberg/pull/37862.

**Comment by sarayourfriend** [comment]
This is awesome. I did a little testing on this branch and everything I tried seemed to work great. @jsnajdr and I spent a considerable amount of time trying to find a solution for this, I'm glad there was something simpler and much more straightforward to use possible! Really nice work. Will be happy to review again once it's converted from a draft to a full PR.

File: packages/compose/src/higher-order/if-condition/index.tsx
Comment (dmsnell) [reply packages/compose/src/higher-order/if-condition/index.tsx]: rebased against new `trunk`, which contains #37862 
thanks again @gziolo !

**Review by sarayourfriend — APPROVED** [review]
Looks great to me! I would love a second opinion from @jsnajdr who tends to see things I miss! But otherwise thanks so much for finding this significant improvement to the usability of this utility.

**Comment by dmsnell** [comment]
Rebased against the fix to `docgen` in #37929.

**Comment by dmsnell** [comment]
cc: @sirbrillig keep an eye out for this in the next release ✅

**Comment by dmsnell** [comment]
Merging for lack of further review. Should anyone notice any problems with this please don't hesitate to ping me or ask for further refinement.

**Comment by noahtallen** [comment]
I'm just now getting around to updating this package in a 3rd party repo. We have this use case:

```js
export const withLocale = createHigherOrderComponent< { locale: string } >( ( InnerComponent ) => {
	return ( props ) => {
		const locale = useLocale();
		return <InnerComponent { ...props } locale={ locale } />;
	};
}, 'withLocale' );
```

Which fails compilation with:

```
packages/i18n-utils/src/locale-context.tsx(99,11): error TS2322: Type 'Omit<InnerProps, "locale"> & { locale: string; children?: ReactNode; }' is not assignable to type 'IntrinsicAttributes & InnerProps & { children?: ReactNode; }'.
  Type 'Omit<InnerProps, "locale"> & { locale: string; children?: ReactNode; }' is not assignable to type 'InnerProps'.
    'Omit<InnerProps, "locale"> & { locale: string; children?: ReactNode; }' is assignable to the constraint of type 'InnerProps', but 'InnerProps' could be instantiated with a different subtype of constraint '{ locale: string; }'.
```

To me, it seems like this super basic use case should be supported without adding, e.g. `@ts-ignore`. Maybe I'm missing something? It seems like `ts-ignore` was used in this PR [to fix the instance ID hook](https://github.com/WordPress/gutenberg/pull/37795/files#diff-8ec5fdf6f5e927f311631615e2f03d0ebda9173f6c74bda3bdc7e38f8f2e55d8R16).

File: packages/compose/src/higher-order/with-safe-timeout/index.tsx
Comment (jsnajdr) [reply packages/compose/src/higher-order/with-safe-timeout/index.tsx:73]: > Why do we need this type assertion?

It's described in this TS issue: https://github.com/microsoft/TypeScript/issues/35858#issuecomment-573909154

Quote:

> TS isn't capable of the higher-order reasoning needed to understand that `Exclude<T, k> & { [k]: T[k] }` is equivalent to T.

If you observe how the type of `props` is `Omit<TProps, ...>` and how the value of local `props` value is constructed with `{ ...this.props, ...}`, it's exactly the same situation.

**Comment by jsnajdr** [comment]
@noahtallen I was able to make `withLocale` work by adding some explicit types:
```ts
type LocaleProps = { locale: string };
export const withLocale = createHigherOrderComponent< LocaleProps >(
  < TProps extends LocaleProps >( InnerComponent: React.ComponentType< TProps > ) => {
    return ( props: Omit< TProps, keyof LocaleProps > ) => {
      const locale = useLocale();
      const innerProps = { locale, ...props } as TProps;
      return <InnerComponent { ...innerProps } />;
    };
  },
  'withLocale'
);
```
Note that Gutenberg's `withSafeTimeout` HOC uses the same technique.

The culprit is the same as described in https://github.com/WordPress/gutenberg/pull/37795#discussion_r862692495. TypeScript infers the `InnerProps` type from the type of `InnerComponent`, but then is unable to figure out that `Omit< InnerProps, 'locale' > & { locale: string }` is equivalent to `InnerProps`. We humans understand that removing the `locale` field and then adding it back again leads to the original `InnerProps` type, but TypeScript doesn't. So there needs to be an explicit `as InnerProps` cast. But to get access to `InnerProps`, I need to declare the entire type of the `mapComponent` function explicitly.

Typing HOCs in TypeScript is incredibly painful, the two concepts are not very compatible. Everything's much simpler when everything is just hooks and functional components.

**Comment by jsnajdr** [comment]
...and here's an easier way to fix `withLocale`, utilizing the `ComponentProps` helper that extracts the `InnerProps` type from `InnerComponent`:
```ts
const innerProps = { locale, ...props } as React.ComponentProps<typeof InnerComponent>;
return <InnerComponent { ...innerProps } />;
```

**Comment by dmsnell** [comment]
Thanks @jsnajdr for the sleuthing. I am hoping too that the newer `in`/`out` covariant/contravariant syntax will let us clean this up nicely. Do you propose we update the types internally in the meantime to somehow fix this?

> Typing HOCs in TypeScript is incredibly painful, the two concepts are not very compatible.

No need to clarify, but I'm curious what mismatch you are talking about. When working on this I did find some situations frustrating in that I thought I should be able to rely on the type inference but couldn't.

If you have a better idea at the fundamental issues maybe that would help us find a better solution?

**Comment by jsnajdr** [comment]
> No need to clarify, but I'm curious what mismatch you are talking about. When working on this I did find some situations frustrating in that I thought I should be able to rely on the type inference but couldn't.

It's just that when typing a HOC, things are never simple. You do something wrong, and suddenly you get a very abstract warning like:
```
Type 'X' is not assignable to type 'Y'. 'X' is assignable to the constraint of type 'Y',
but 'Y' could be instantiated with a different subtype of constraint 'Z'.
```
One needs to do some considerable mathematical thinking to unpack this.

Because everything is hard, the types are rarely pleasant to work with, like having inference where one would expect it, or are just plain wrong. Consider this usage of `ifCondition`:
```ts
type PropsOne = { name: string };
type PropsTwo = { count: number };

const Foo = ( { name }: PropsOne ) => <div>{ name }</div>;
const predicate = ( { count }: PropsTwo ) => count > 0;

const IfFoo = ifCondition( predicate )( Foo );
```
There are types defined, typechecking passes successfully, but there is an obvious type mismatch: the `predicate` argument must be the of same type as the props of `Foo`, but they are completely different and TypeScript didn't catch it.

When I tried to fix `ifCondition` to ensure that the types of `predicate` and `Foo` are compatible, and to make `ifCondition( predicate )` accept only components whose props match the `predicate` (it can't accept just any component, like other HOCs!), I got this:
```
Argument of type '<InnerProps extends {}>(WrappedComponent: ComponentType<TProps>) => (props: TProps) =>
Element | null' is not assignable to parameter of type 'HigherOrderComponent<{}>'. Types of parameters
'WrappedComponent' and 'Inner' are incompatible. Type 'ComponentType<InnerProps>' is not assignable to type
'ComponentType<TProps>'. Type 'ComponentClass<InnerProps, any>' is not assignable to type 'ComponentType<TProps>'.
Type 'ComponentClass<InnerProps, any>' is not assignable to type 'ComponentClass<TProps, any>'. Types of property
'propTypes' are incompatible. Type 'WeakValidationMap<InnerProps> | undefined' is not assignable to type
'WeakValidationMap<TProps> | undefined'. Type 'WeakValidationMap<InnerProps>' is not assignable to type
'WeakValidationMap<TProps>'. Type 'keyof TProps' is not assignable to type 'keyof InnerProps'. Type 'string | number |
symbol' is not assignable to type 'keyof InnerProps'. Type 'string' is not assignable to type 'keyof InnerProps'. Type 'string'
is not assignable to type 'never'. Type 'keyof TProps' is not assignable to type 'never'. Type 'string | number | symbol' is not
assignable to type 'never'. Type 'string' is not assignable to type 'never'.
```
How to proceed from here? It's no longer fun at all.

Also, I don't know what was the motivation for the changes you did in this PR, you merely wrote that "existing types in `createHigherOrderComponent<TOuter, TInner>` are somewhat
_surprising_". But it seems to me that declaring the `TOuter` and `TInner` types was more correct than declaring the "difference".

A HOC is a function that transforms one component type to another, and their props don't need to be related at all. The case where new props are injected and old ones are passed is merely a special case, although the most frequent one. It's like when the type of `Array<A>.map<B>` is `( value: A ) => B`, with no relation between `A` and `B`.

**Comment by dmsnell** [comment]
Thanks for clarifying. I'll work on this and see if we can't make those errors clearer. I don't know if this feels like an inherent mismatch though that would be better with hooks, unless the claim is that with hooks we just don't have to worry about wrapped props. It looks like you have pointed out a flaw in my assumptions in #37795

Noted in that PR I wasn't fully confident in the solution, but I did let it sit open for a few weeks to try and find review on it at the time. The motivation was some confusion around how to properly type the HOC. IIRC I was primarily considering the prop-injecting HOC pattern and wanted to warn people if they were trying to supply a prop that would ultimately be ignored. In Gutenberg I see counter-examples such as the color-providing HOC.

So again I think you're right in pointing out that the types were more correct before. There may still be a way to revert to that while removing the additional type wrappers like `PropInjectingHigherOrderComponent` that might have muddied the water.

**Comment by jsnajdr** [comment]
> I was primarily considering the prop-injecting HOC pattern and wanted to warn people if they were trying to supply a prop that would ultimately be ignored. In Gutenberg I see counter-examples such as the color-providing HOC.

Is there a specific example of a HOC that had troubles with the old types and got better with the new types? I'd like to understand the issues better.

At this moment I feel that we want to revert this PR to the old state, but not _completely_? There are some use cases that this PR actually improved, and we want to do some kind of synthesis?

**Comment by dmsnell** [comment]
Looking back on some issues with Calypso code what I believe was happening is that the type of `InnerProps` was enforced for all invocations of the produced higher-order-component when in fact the produced higher-order-component shouldn't demand the same input props to the inner component.

https://github.com/WordPress/gutenberg/pull/37795/files#diff-05e2e8846bd708cfa15b312de800aa48fdab481b284063a3e7de6d19fd245717L39-L44

Or put more minimally, we had the kind of problem of using this…

```ts
interface ToBinary<T> {
    (input: T): Binary;
}
```

…when we want this instead…

```ts
interface ToBinary {
    <T>(input: T): Binary;
}
```

Screen Shot 2022-05-10 at 1 50 07 PM


You can see in this patch where I was trying to do that, but added the unnecessary and inaccurate constraint that `InnerProps` extends `HOCProps`

**Pull request merged**
